nefertari/events.py

```diff
@@ -0,0 +1,198 @@
+        'only. Should be used to read data only. Changes to the instance '
+        'may result in database data inconsistency.')
+
+
+@implementer(IRequestEvent)
+class RequestEvent(object):
+    """ Nefertari request event. """
+    def __init__(self, model, view,
+                 fields=None, field=None, instance=None):
+        self.model = model
+        self.view = view
+        self.fields = fields
+        self.field = field
+        self.instance = instance
+
+
+# 'Before' events
+
+class before_index(RequestEvent):
+    pass
+
+
+class before_show(RequestEvent):
+    pass
+
+
+class before_create(RequestEvent):
+    pass
+
+
+class before_update(RequestEvent):
+    pass
+
+
+class before_replace(RequestEvent):
+    pass
+
+
+class before_delete(RequestEvent):
+    pass
+
+
+class before_update_many(RequestEvent):
+    pass
+
+
+class before_delete_many(RequestEvent):
+    pass
+
+
+class before_item_options(RequestEvent):
+    pass
+
+
+class before_collection_options(RequestEvent):
+    pass
+
+
+# 'After' events
+
+class after_index(RequestEvent):
+    pass
+
+
+class after_show(RequestEvent):
+    pass
+
+
+class after_create(RequestEvent):
+    pass
+
+
+class after_update(RequestEvent):
+    pass
+
+
+class after_replace(RequestEvent):
+    pass
+
+
+class after_delete(RequestEvent):
+    pass
+
+
+class after_update_many(RequestEvent):
+    pass
+
+
+class after_delete_many(RequestEvent):
+    pass
+
+
+class after_item_options(RequestEvent):
+    pass
+
+
+class after_collection_options(RequestEvent):
+    pass
+
+
+# Subscriber predicates
+
+class ModelClassIs(object):
+    """ Subscriber predicate to check event.model is the right model. """
+
+    def __init__(self, model, config):
+        """
+        :param model: Model class
+        """
+        self.model = model
+
+    def text(self):
+        return 'Model class is %s' % (self.model,)
+
+    phash = text
+
+    def __call__(self, event):
+        """ Check whether one of following is true:
+
+        * event.model is the same class as self.model
+        * event.model is subclass of self.model
+        """
+        return (event.model is self.model or
+                issubclass(event.model, self.model))
+
+
+class FieldIsChanged(object):
+    """ Subscriber predicate to check particular field is changed. """
+
+    def __init__(self, field, config):
+        self.field = field
+
+    def text(self):
+        return 'Field `%s` is changed' % (self.field,)
+
+    phash = text
+
+    def __call__(self, event):
+        if self.field in event.fields:
+            event.field = event.fields[self.field]
+            return True
+        return False
+
+
+@contextmanager
+def trigger_events(view_obj):
+    """ Trigger before_ and after_ CRUD events.
+
+    :param view_obj: Instance of nefertari.view.BaseView subclass created
+        by nefertari.view.ViewMapper.
+    """
+    from nefertari.utils import FieldData
+    _globals = globals()
```

Comment:
There is bad code smell here :-)

You should really use a lookup table mapping actions to event classes.
You could even add an `action` attribute to event classes and build the lookup table from that.